text/0000-cargo-templates.md

```diff
@@ -0,0 +1,275 @@
+
+The community has also created two crates that solve similar problems:
+
+* [cargo-generate](https://crates.io/crates/cargo-generate): a developer tool to help you get up and running quickly with a new Rust project by leveraging a pre-existing git repository as a template
+* [kickstart](https://github.com/Keats/kickstart): created by @Keats described as "A scaffolding tool to get new projects up and running quickly"
+* [cookiecutter](https://github.com/cookiecutter/cookiecutter) : not a Rust project, but "A command-line utility that creates projects from cookiecutters (project templates)"
+
+In addition, there was a lot of discussion both from the Cargo team and the community on [this issue](https://github.com/rust-lang/cargo/issues/3506). It has been decided that there [still remains disagreement](https://github.com/rust-lang/cargo/pull/8029#issuecomment-604756599) among both the Cargo team and the community on how to solve this, hence why this RFC seems to be the logical next step.
+
+# What lessons can we learn from what other communities have done here?
+
+There are pros and cons to having templates or some type of template ecosystem. I think the biggest question is **who will maintain them?** It may not be directly related, but it's an important point to consider should templates be added to `cargo new` .
+
+If we return to our examples from the JavaScript industry, there are two that stick out:
+
+## Community maintained templates
+
+In the Gatsby.js community, there are +[300 starters](https://www.gatsbyjs.org/starters/?v=2). Only a select few are maintained by the Gatsby.js core team. The rest are added by community members.
+
+### Pros
+
+* Community members can contribute
+* There is a wilder selection of options
+
+### Cons
+
+* Members can abandon their starters which can negatively impact the community
+
+## Core team maintained templates
+
+Revisiting the `create-react-app` and the `create-next-app` templates, those are maintained by core team members (i.e. a select group of individuals). There are some variations to the templates (i.e. regular vs. TypeScript).
+
+### Pros
+
+* Higher-quality
+* Creates a “standard”
+* Reliable
+
+### Cons
+
+* Requires dedicated maintainers
+* Less community involvement (beyond direct contributions)
+
+# Unresolved questions
+
+Some questions that may require further discussion depending on how this RFC goes.
+
+Questions that fall in the scope of this RFC:
+
+## What is a template ?
+
+> A template is a directory that lives inside `crate/templates/`. It must include the following
+> 
+> The `Cargo.toml` of a template must include the following text (or some other placeholder) literally at some point, which will be replaced wholesale by the template engine. - @CAD97
+
+```
+[package]
+name = ""
+authors = []
+```
+
```

Comment:
@ckaran I feel like we already addressed that in the Pre-RFC, no?

Reply:
Yes, we worked it out in the Pre-RFC discussion, but I don't see that copied into here.  Are you going to capture that discussion in the RFC, so that everyone sees precisely what is being proposed?

Reply:
> Yes, we worked it out in the Pre-RFC discussion, but I don't see that copied into here.

I was under the impression that _I have_ copied that into here. 

> Are you going to capture that discussion in the RFC, so that everyone sees precisely what is being proposed?

However, I'm getting the idea that you feel like something is missing from this "What is a template?" question.  What else do you feel included?

I tried my best to take the summarized pieces from the discussion and include them in the RFC. 

I don't want to copy everything over though (unless multiple people agree that's necessary). I feel like the RFC captures the essence of our discussions, and if people want to read more, they can visit the links to the Pre-RFCs.

Reply:
It's probably a difference in view point; when I see an RFC like this, I'm kind of expecting it to be more like a spec that I can then code to.  The RFC as it currently stands leaves a number of questions unanswered:

- Is this a pure text substitution engine, or does it perform macro expansion and arbitrary code execution?  
- If a template represents a complete project, does that mean that you have a complete directory that is copied over somewhere, and then modified, or is the template a single file that gets expanded?
- How do you indicate what parts of the template are to be copied over, and what parts are to be replaced?  How do you document this such that the documentation of the template doesn't get copied over?
- Are templates for entire projects, or can they be for portions of a project (e.g. a project template and a file template)?

I know that we talked about all of this in the pre-RFC discussion, and all the details got hammered out, but the RFC really needs to capture all of that without any ambiguity.  After all, the RFC will be the primary reference for this feature.  When people ask 'is it done?  Does it work to spec?', they'll read the RFC to make that decision.  Right now there are too many open questions within the RFC itself to answer those types of questions.

Reply:
I'm going to resolve this conversation in favor of moving it to the main comment thread